cdc/owner/async_sink.go

```diff
@@ -46,82 +47,99 @@ type AsyncSink interface {
 	ddlFinishedTs model.Ts
 	ddlSentTs     model.Ts
 
+	ddlCh chan *model.DDLEvent
+	errCh chan error
+
+	sink            sink.Sink
+	sinkInitHandler asyncSinkInitHandler
+
 	cancel context.CancelFunc
-	wg     sync.WaitGroup
-	errCh  chan error
 }
 
-func newAsyncSink(ctx cdcContext.Context) (AsyncSink, error) {
-	ctx, cancel := cdcContext.WithCancel(ctx)
-	changefeedID := ctx.ChangefeedVars().ID
-	changefeedInfo := ctx.ChangefeedVars().Info
-	filter, err := filter.NewFilter(changefeedInfo.Config)
+func newAsyncSink() AsyncSink {
+	return &asyncSinkImpl{
+		ddlCh:           make(chan *model.DDLEvent, 1),
+		errCh:           make(chan error, defaultErrChSize),
+		sinkInitHandler: asyncSinkInitializer,
+		cancel:          func() {},
+	}
+}
+
+type asyncSinkInitHandler func(ctx cdcContext.Context, a *asyncSinkImpl) error
+
+func asyncSinkInitializer(ctx cdcContext.Context, a *asyncSinkImpl) error {
+	var (
+		id   = ctx.ChangefeedVars().ID
+		info = ctx.ChangefeedVars().Info
+	)
+
+	filter, err := filter.NewFilter(info.Config)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return errors.Trace(err)
 	}
-	errCh := make(chan error, defaultErrChSize)
-	s, err := sink.New(ctx, changefeedID, changefeedInfo.SinkURI, filter, changefeedInfo.Config, changefeedInfo.Opts, errCh)
+
+	s, err := sink.New(ctx, id, info.SinkURI, filter, info.Config, info.Opts, a.errCh)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return errors.Trace(err)
 	}
-	asyncSink := &asyncSinkImpl{
-		sink:   s,
-		ddlCh:  make(chan *model.DDLEvent, 1),
-		errCh:  errCh,
-		cancel: cancel,
+	a.sink = s
+
+	if !info.SyncPointEnabled {
+		return nil
 	}
-	if changefeedInfo.SyncPointEnabled {
-		asyncSink.syncpointStore, err = sink.NewSyncpointStore(ctx, changefeedID, changefeedInfo.SinkURI)
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-		if err := asyncSink.syncpointStore.CreateSynctable(ctx); err != nil {
-			return nil, errors.Trace(err)
-		}
+	syncPointStore, err := sink.NewSyncpointStore(ctx, id, info.SinkURI)
+	if err != nil {
+		return errors.Trace(err)
 	}
-	asyncSink.wg.Add(1)
-	go asyncSink.run(ctx)
-	return asyncSink, nil
+	a.syncPointStore = syncPointStore
+
+	if err := a.syncPointStore.CreateSynctable(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
 }
 
-func (s *asyncSinkImpl) run(ctx cdcContext.Context) {
-	defer s.wg.Done()
+func (s *asyncSinkImpl) Run(ctx cdcContext.Context) error {
+	ctx, cancel := cdcContext.WithCancel(ctx)
+	s.cancel = cancel
+
+	start := time.Now()
+	if err := s.sinkInitHandler(ctx, s); err != nil {
+		return errors.Trace(err)
+	}
+
+	log.Info("async sink initialized, start processing...", zap.Duration("elapsed", time.Since(start)))
+
 	// TODO make the tick duration configurable
 	ticker := time.NewTicker(time.Second)
 	defer ticker.Stop()
 	var lastCheckpointTs model.Ts
 	for {
 		select {
 		case <-ctx.Done():
-			return
+			return nil
 		case err := <-s.errCh:
-			ctx.Throw(err)
-			return
+			return err
```

Comment:
```suggestion
			return errors.Trace(err)
```

Also, could you add some comments to explain why not throw error?

Reply:
since it's thrown at the outside.